pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificaterequests
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	"github.com/kr/pretty"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/validation"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	errorCertificateNotExists = "CertNotExists"
+	errorCertificateParse     = "CertParseError"
+
+	errorIssuerNotFound = "IssuerNotFound"
+	errorIssuerInit     = "IssuerInitError"
+
+	successCertificateIssued = "CertIssued"
+)
+
+func (c *Controller) Sync(ctx context.Context, cr *v1alpha1.CertificateRequest) (err error) {
+	c.metrics.IncrementSyncCallCount(ControllerName)
+
+	log := logf.FromContext(ctx)
+	dbg := log.V(logf.DebugLevel)
+
+	crCopy := cr.DeepCopy()
+	defer func() {
+		if _, saveErr := c.updateCertificateRequestStatus(ctx, cr, crCopy); saveErr != nil {
+			err = utilerrors.NewAggregate([]error{saveErr, err})
+		}
+	}()
+
+	dbg.Info("Fetching existing certificate signing request and certificate from certificate request",
+		"name", crCopy.ObjectMeta.Name)
+
+	issuerObj, err := c.helper.GetGenericIssuer(crCopy.Spec.IssuerRef, crCopy.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		c.log.Error(err,
+			"issuer-name", crCopy.Spec.IssuerRef.Name,
+			"issuer-kind", crCopy.Spec.IssuerRef.Kind)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	issuerType, err := apiutil.NameForIssuer(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		c.log.Error(err,
+			"issuer-name", issuerObj.GetName())
+		return nil
+	}
+
+	// This CertificateRequest is not meant for us, ignore
+	if issuerType != c.issuerType {
+		c.log.V(5).Info("issuer reference type does not match resource kind, ignoring",
+			"certificaterequest-issuer-type", issuerType,
+			"issuer-type", c.issuerType)
+		return nil
+	}
+
+	el := validation.ValidateCertificateRequest(crCopy)
```

Comment:
I think this validation can (and probably should?) happen much earlier in the processing flow. We open ourselves up to acting on malformed resources if we wait until later in the control flow to perform validation. It should be as early as possible IMO.

Reply:
The issue with this is that we will then have n number of controllers all spamming the logs telling us the spec is wrong. Should we not leave to the one and only controller that 'owns' the resource?

Reply:
The problem with that is that if it doesn't belong to any controller then there will never be an error if webhook is disabled. :woman_shrugging:

Reply:
This is a tricky one, because an invalid resource could be invalid because it is e.g. missing the `issuerRef`, in which case this BadConfig event would never be fired.

Perhaps doing this earlier and updating/adding a status condition for BadConfig would be best? Then, all controllers would attempt the validation and attempt to update the resource, and only one will succeed due to the resourceVersion difference. It'd at least avoid having N events logged by each controller.